src/ray/core_worker/core_worker.cc

```diff
@@ -2204,6 +2206,52 @@ Status CoreWorker::SealReturnObject(const ObjectID &return_id,
   return status;
 }
 
+bool CoreWorker::PinExistingReturnObject(const ObjectID &return_id,
+                                         std::shared_ptr<RayObject> *return_object) {
+  // TODO(swang): It would be better to evict the existing copy instad of
+  // reusing it, to make sure it's consistent with the task re-execution.
+  absl::flat_hash_map<ObjectID, std::shared_ptr<RayObject>> result_map;
+  bool got_exception;
+  rpc::Address owner_address(worker_context_.GetCurrentTask()->CallerAddress());
+
+  // Temporarily set the return object's owner's address. This is needed to retrieve the
+  // value from plasma.
+  reference_counter_->AddLocalReference(return_id, "<temporary (pin return object)>");
+  reference_counter_->AddBorrowedObject(return_id, ObjectID::Nil(), owner_address);
+
+  auto status = plasma_store_provider_->Get({return_id}, 0, worker_context_, &result_map,
+                                            &got_exception);
+  // Remove the temporary ref.
+  RemoveLocalReference(return_id);
+
+  if (result_map.count(return_id)) {
+    *return_object = std::move(result_map[return_id]);
+    RAY_LOG(DEBUG) << "Pinning existing return object " << return_id
+                   << " owned by worker "
+                   << WorkerID::FromBinary(owner_address.worker_id());
+    // Keep the object in scope until it's been pinned.
+    std::shared_ptr<RayObject> pinned_return_object = *return_object;
+    local_raylet_client_->PinObjectIDs(
+        owner_address, {return_id},
+        [return_id, pinned_return_object](const Status &status,
+                                          const rpc::PinObjectIDsReply &reply) {
+          if (!status.ok()) {
+            RAY_LOG(INFO) << "Failed to pin existing copy of the task return object "
```

Comment:
I don't think it's really supposed to happen, but I think it's better to rely on higher-level failure handling (e.g., fetch timeout) than mark the object as lost.

Reply:
Hmm can you comment that how the error is handled? Also, this means although the function itself returns "success", it is possible it could fail to pin it right? Can you also comment about this behavior in the docstring of this method? (e.g., Although the it returns success, the object may not be pinned if X happens or something like that)

Reply:
Will do. In general, the owner should be handling the failure, not the worker.

This behavior is the same as the usual return objects path, btw.